focus/operations/src/maintenance/mod.rs

```diff
@@ -328,22 +346,106 @@ impl Runner<'_> {
         info!(?time_period, ?path, "running tasks");
         set_default_git_maintenance_config(path)?;
+
+        let maint_result_metric = "maint_result".to_string();
         match self.run_maint(time_period, path) {
             Ok(MaintResult::Success(status)) => {
                 if status.success() {
                     debug!(?time_period, ?path, "completed maintenance",)
                 } else {
                     warn!(?path, exit_status = ?status, "maintenance failed");
+                    let mut maint_custom_map = HashMap::new();
+                    maint_custom_map.insert(
+                        maint_result_metric,
+                        self.maint_exitstatus_metric_helper(status),
+                    );
+                    self.add_ti_invocation_message(&maint_custom_map);
                 }
             }
-            Ok(MaintResult::LockFailed) => warn!(?path, "failed to acquire lock"),
+            Ok(MaintResult::LockFailed) => {
+                warn!(?path, "failed to acquire lock");
+                let maint_custom_map =
+                    HashMap::from([(maint_result_metric, "lock_failed".to_string())]);
+                self.add_ti_invocation_message(&maint_custom_map);
+            }
             Err(e) => {
-                warn!(?path, ?e, "failed runing git-maintenance");
+                warn!(?path, ?e, "failed running git-maintenance");
+                let maint_custom_map =
+                    HashMap::from([(maint_result_metric, "git_error".to_string())]);
+                self.add_ti_invocation_message(&maint_custom_map);
             }
         }
         Ok(())
     }
 
+    fn add_ti_invocation_message(&self, maint_custom_map: &HashMap<String, String>) {
+        let ti_client = self.app.tool_insights_client();
+        ti_client.get_inner().add_invocation_message(
+            SystemTime::now(),
+            None,
+            Some(maint_custom_map),
+        );
+    }
+
+    fn maint_exitstatus_metric_helper(&self, exit_status: ExitStatus) -> String {
+        match exit_status.code() {
+            Some(code) => code.to_string(),
+            None => "signal_terminated".to_string(),
+        }
+    }
+
+    fn sync_status_metric_helper(&self, sync_status: crate::sync::SyncStatus) -> String {
+        match sync_status {
+            crate::sync::SyncStatus::Success => "success".to_string(),
+            crate::sync::SyncStatus::SkippedSyncPointUnchanged => "skipped_unchanged".to_string(),
+            crate::sync::SyncStatus::SkippedSyncPointDifferenceIrrelevant => {
+                "skipped_irrelevant".to_string()
+            }
+            crate::sync::SyncStatus::SkippedPreemptiveSyncDisabled => {
+                "skipped_disabled".to_string()
+            }
+            crate::sync::SyncStatus::SkippedPreemptiveSyncCancelledByActivity => {
+                "skipped_activity".to_string()
+            }
+        }
+    }
```

Comment:
For functions like these which don't refer to `self` and just does pure computation, it's probably best to make them free functions at the top-level.

Reply:
Fixed. Moved those functions to the top level.